src/Microsoft.OData.Client/DataServiceExtensions.cs

```diff
@@ -0,0 +1,173 @@
+
+            transientResult = (IEnumerable<TTarget>)distinctMethod.MakeGenericMethod(
+                typeof(TTarget)).Invoke(null, new object[] { transientResult });
+
+            // Method: Count<TSource>(IEnumerable<TSource>)
+            MethodInfo countMethod = GetCountMethod(typeof(Enumerable), typeof(IEnumerable<>));
+
+            return (int)countMethod.MakeGenericMethod(
+                typeof(TTarget)).Invoke(null, new object[] { transientResult });
+        }
+
+        /// <summary>
+        /// Returns Select method defined in Queryable type.
+        /// </summary>
+        private static MethodInfo GetQueryableSelectMethod()
+        {
+            return typeof(Queryable).GetMethods(BindingFlags.Static | BindingFlags.Public)
+                .Where(d1 => d1.Name.Equals("Select", StringComparison.Ordinal))
+                .Select(d2 => new { Method = d2, Parameters = d2.GetParameters() })
+                .Where(d3 => d3.Parameters.Length.Equals(2)
+                    && d3.Parameters[0].ParameterType.IsGenericType
+                    && d3.Parameters[0].ParameterType.GetGenericTypeDefinition().Equals(typeof(IQueryable<>))
+                    && d3.Parameters[1].ParameterType.IsGenericType
+                    && d3.Parameters[1].ParameterType.GetGenericTypeDefinition().Equals(typeof(Expression<>)))
+                .Select(d4 => new { d4.Method, SelectorArguments = d4.Parameters[1].ParameterType.GetGenericArguments() })
+                .Where(d5 => d5.SelectorArguments.Length.Equals(1)
+                    && d5.SelectorArguments[0].IsGenericType
+                    && d5.SelectorArguments[0].GetGenericTypeDefinition().Equals(typeof(Func<,>)))
+                .Select(d6 => d6.Method).Single();
+        }
+
+        /// <summary>
+        /// Returns Select method defined in Enumerable type.
+        /// </summary>
+        private static MethodInfo GetEnumerableSelectMethod()
+        {
+            return typeof(Enumerable).GetMethods(BindingFlags.Static | BindingFlags.Public)
+                .Where(d1 => d1.Name.Equals("Select", StringComparison.Ordinal))
+                .Select(d2 => new { Method = d2, Parameters = d2.GetParameters() })
+                .Where(d3 => d3.Parameters.Length.Equals(2)
+                    && d3.Parameters[0].ParameterType.IsGenericType
+                    && d3.Parameters[0].ParameterType.GetGenericTypeDefinition().Equals(typeof(IEnumerable<>))
+                    && d3.Parameters[1].ParameterType.IsGenericType
+                    && d3.Parameters[1].ParameterType.GetGenericTypeDefinition().Equals(typeof(Func<,>)))
+                .Select(d6 => d6.Method).Single();
+        }
+
+        /// <summary>
+        /// Returns Distinct method defined in <paramref name="declaringType"/>.
+        /// </summary>
+        private static MethodInfo GetDistinctMethod(Type declaringType, Type sourceType)

```

Comment:
>GetDistinctMethod [](start = 34, length = 17)

should be static (cache) it?

Reply:
@xuzhg Used `SimpleLazy` class, let me know if its a good fit